File: modules/angular2/src/router/lifecycle_annotations_impl.ts
import {CONST, CONST_EXPR} from 'angular2/src/facade/lang';

@CONST()
export class RouteLifecycleHook {
  constructor(public name: string) {}
}

@CONST()
export class CanActivate {
  constructor(public fn: Function) {}
}

export const canReuse: RouteLifecycleHook = CONST_EXPR(new RouteLifecycleHook("canReuse"));
export const canDeactivate: RouteLifecycleHook =
    CONST_EXPR(new RouteLifecycleHook("canDeactivate"));
export const onActivate: RouteLifecycleHook = CONST_EXPR(new RouteLifecycleHook("onActivate"));
export const onReuse: RouteLifecycleHook = CONST_EXPR(new RouteLifecycleHook("onReuse"));
export const onDeactivate: RouteLifecycleHook = CONST_EXPR(new RouteLifecycleHook("onDeactivate"));

File: lib/src/router/lifecycle_annotations_impl.dart
library angular2.src.router.lifecycle_annotations_impl;

class RouteLifecycleHook {
  final String name;
  const RouteLifecycleHook(this.name);
}

class CanActivate {
  final Function fn;
  const CanActivate(this.fn);
}

const RouteLifecycleHook canReuse = const RouteLifecycleHook("canReuse");
const RouteLifecycleHook canDeactivate =
    const RouteLifecycleHook("canDeactivate");
const RouteLifecycleHook onActivate = const RouteLifecycleHook("onActivate");
const RouteLifecycleHook onReuse = const RouteLifecycleHook("onReuse");
const RouteLifecycleHook onDeactivate =
    const RouteLifecycleHook("onDeactivate");
